2018-10/workshop-two/workshop-two.md

```diff
@@ -0,0 +1,386 @@
+    fmt.Fprintf(w, "Hello World")
+}
+
+func main() {
+    //...
+    locationHandler := locationHandler{}
+    http.Handle("/", locationHandler)
+    if err := http.ListenAndServe(":"+*port, nil); err != nil {
+        fmt.Printf("failed to start server: %v", err)
+        os.Exit(1)
+    }
+}
+```
+
+```bash
+go run . -secret=abc123 # secret=abc123 gin .
+```
+
+---
+
+# Weather (aka Location) Handler
+
+http://localhost:3000?lat=53.5&lng=-113.3
+
+---
+
+# Querystring params
+
+```go
+q := r.URL.Query()
+lat, err := strconv.ParseFloat(q.Get("lat"), 10)
+if err != nil {
+    http.Error(w, "Invalid lat querystring param", http.StatusBadRequest)
+    return
+}
+```
+
+## Exercise
+
+Parse out the longitude value as well, then test it via the browser
+
+---
+
+# Get the weather 
+
+```go
+var darkskySvc *darksky.DarkSky
+func main() {
+    // ...
+    darkskySvc = darksky.New(secret)
+    // http.Handle("/", ...)
+}
+```
+
+```go
+import "github.com/edmontongo/darksky"
+func (lh locationHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+    //...
+    weather, err := darkskySvc.Forecast(darksky.Location{ Lat: lat, Long: lng })
+    if err != nil {
+        http.Error(w, "failed to fetch weather details", http.StatusInternalServerError)
+        return
+    }
+}
+```
+
+---
+
+# Content-Type 
+
+Add the code to the locationHandler
+
+```go
+accept := r.Header.Get("Accept")
+contentType := "application/json"
+if strings.Contains(accept, "text/html") {
+    contentType = "text/html"
+}
+w.Header().Add("Content-Type", contentType)
+
+switch contentType {
+case "text/html":
+    // render html here
+case "application/json":
+    // return json
+}
+```
+
+---
+
+# JSON Marshaling
+
+```go
+import "encoding/json"
+```
+
+```go
+b, err := json.Marshal(weather)
+if err != nil {
+    http.Error(w, "Oops something went wrong", http.StatusInternalServerError)
+    return
+}
+w.Write(b)
+```
+
+---
+
+# Html templates
+
+```go
+t, err := template.New("app.html").
+    Funcs(template.FuncMap{
+        "toCelsius": func(tempF float64) string {
+            degC := math.Round((tempF - 32) * 5 / 9)
+            return fmt.Sprintf("%.0f°C", degC)
+        },
+    }).
+    ParseFiles(
+        "templates/app.html",
+        "views/location.html",
+    )
+if err != nil { return }  // handle these errors
+if err = t.Execute(w, weather); err != nil { return }
+```
+
+---
+
+Add the HTML rendering code
+
+```html
+<!-- templates/app.html -->
+<!DOCTYPE html>
+<html lang="en">
+<head><meta charset="UTF-8"></head>
+<body>
+{{block "content" .}}
+    This text will render if no `content` block is defined
+{{end}}
+</body></html>
+```
+
+```html
+<!-- views/location.html -->
+{{define "content"}}
+<h1>{{.Currently.Summary}}</h1>
+<div>{{.Currently.Temperature | toCelsius}}</div>
+{{end}}
+```
+
+
+---
+
+# Refactor
+
+1. Remove global darkskySvc
+2. Cache templates (won't cover this, but checkout the code on github)
```

Comment:
Maybe we could add this server code to the darksky repo for people to play with -- perhaps not until after the meetup though, given that it's already Monday.